File: backend/app/services/import_service.py
"""Orchestrates an import: parse result -> reconcile -> classify -> persist, with counts.

This is the single funnel every source flows through, so the reconciliation/classification
behaviour is identical regardless of which parser produced the holdings.
"""

from __future__ import annotations

import json

from sqlalchemy import delete
from sqlalchemy.orm import Session

from app.db.models import Account, Holding, ImportBatch, Instrument, Transaction
from app.domain.taxonomy import AssetClass, Source
from app.parsers.base import ParsedHolding, ParseResult
from app.reconcile.reconciler import (
    find_or_create_account,
    find_or_create_instrument,
    upsert_holding,
)
from app.services.classification_service import classify_instrument, load_override_index


def _persist_transactions(
    db: Session, account: Account, instrument: Instrument, h: ParsedHolding, import_id: int
) -> None:
    """Store dated cashflows for XIRR. Idempotent: re-importing the same source for this
    instrument+account replaces the prior set rather than duplicating it."""
    if not h.transactions:
        return
    db.execute(
        delete(Transaction).where(
            Transaction.instrument_id == instrument.id,
            Transaction.account_id == account.id,
            Transaction.source == h.source.value,
        )
    )
    for t in h.transactions:
        db.add(Transaction(
            instrument_id=instrument.id,
            account_id=account.id,
            date=t.date,
            kind=t.kind,
            units=t.units,
            amount=t.amount,
            price=t.price,
            folio=h.folio,
            source=h.source.value,
            import_id=import_id,
        ))


def _prune_unbacked_tradebook_holdings(db: Session) -> int:
    """Remove tradebook-derived positions that a priced snapshot doesn't confirm.

    A tradebook's "net open position" is only reliable with complete trade history; when a real
    holdings snapshot (Zerodha holdings / CAS) exists for an account, IT is the ground truth for
    what's currently held. Unconfirmed tradebook positions (sold, transferred out, or skewed by a
    split/bonus) must not inflate net worth. Tradebook-only accounts keep their positions.
    """
    priced_accounts = {
        h.account_id for h in db.query(Holding).filter(Holding.current_value.isnot(None)).all()
    }
    if not priced_accounts:
        return 0
    unbacked = db.query(Holding).filter(
        Holding.source == Source.ZERODHA_TRADEBOOK.value,
        Holding.current_value.is_(None),
        Holding.account_id.in_(priced_accounts),
    ).all()
    for h in unbacked:
        db.delete(h)
    return len(unbacked)


def process_parse_result(db: Session, result: ParseResult) -> ImportBatch:
    batch = ImportBatch(
        source=result.source.value,
        file_name=result.file_name,
        status="completed" if not any(d.level == "error" for d in result.diagnostics) else "failed",
        count_parsed=len(result.holdings),
    )
    db.add(batch)
    db.flush()

    index = load_override_index(db)
    imported = merged = duplicate = unclassified = 0
    skipped = sum(1 for d in result.diagnostics if "Skipped" in d.message)

    for h in result.holdings:
        account = find_or_create_account(db, h)
        instrument, existed = find_or_create_instrument(db, h)
        outcome = upsert_holding(db, account, instrument, h, import_id=batch.id)

        if outcome == "duplicate":
            duplicate += 1
        else:
            imported += 1
            if existed:
                merged += 1  # new position reconciled onto a pre-existing instrument

        _persist_transactions(db, account, instrument, h, import_id=batch.id)

        classification = classify_instrument(db, instrument, index)
        if classification.asset_class == AssetClass.UNCLASSIFIED.value:
            unclassified += 1

    batch.count_imported = imported
    batch.count_merged = merged
    batch.count_duplicate = duplicate
    batch.count_skipped = skipped
    batch.count_unclassified = unclassified

    diagnostics = [d.model_dump() for d in result.diagnostics]
    db.flush()  # ensure all just-added holdings are visible to the prune query below
    pruned = _prune_unbacked_tradebook_holdings(db)
    if pruned > 0:
        diagnostics.append({
            "level": "info",
            "message": (
                f"Excluded {pruned} tradebook position(s) not present in your current holdings — "
                "likely sold, transferred out, or affected by a split/bonus. Their trades are kept "
                "for history; they don't count toward net worth."
            ),
            "context": None,
        })
    batch.diagnostics = json.dumps(diagnostics)

    db.flush()
    return batch

File: frontend/src/engine/pipeline/importService.ts
// Import orchestration — TS mirror of backend/app/services/import_service.py.
// parse result → reconcile → persist transactions → classify → counts.

import { Store } from "../store/store";
import { ImportBatch } from "../store/model";
import { AssetClass, Source } from "../taxonomy";
import { ParseResult, ParsedHolding } from "../parse/types";
import { findOrCreateAccount, findOrCreateInstrument, upsertHolding } from "./reconcile";
import { classifyInstrument, loadOverrideIndex } from "./classifyService";

/**
 * Remove tradebook-derived positions that a priced snapshot doesn't confirm.
 * A tradebook's "net open position" is only reliable with complete trade history; when a real
 * holdings snapshot (Zerodha holdings / CAS) exists for an account, IT is the ground truth for
 * what's currently held. Unconfirmed tradebook positions (sold, transferred out, or skewed by a
 * split/bonus) must not be added to net worth. Tradebook-only accounts keep their positions.
 */
export function pruneUnbackedTradebookHoldings(store: Store): number {
  const pricedAccounts = new Set(store.holdings.filter((h) => h.current_value != null).map((h) => h.account_id));
  const before = store.holdings.length;
  store.holdings = store.holdings.filter(
    (h) => !(h.source === Source.ZERODHA_TRADEBOOK && h.current_value == null && pricedAccounts.has(h.account_id)),
  );
  return before - store.holdings.length;
}

function persistTransactions(store: Store, accountId: number, instrumentId: number, h: ParsedHolding, importId: number): void {
  if (!h.transactions.length) return;
  // Idempotent: re-importing the same source for this instrument+account replaces the set.
  store.transactions = store.transactions.filter(
    (t) => !(t.instrument_id === instrumentId && t.account_id === accountId && t.source === h.source)
  );
  for (const t of h.transactions) {
    store.transactions.push({
      id: store.nextId("transactions"),
      instrument_id: instrumentId,
      account_id: accountId,
      date: t.date,
      kind: t.kind,
      units: t.units ?? null,
      amount: t.amount,
      price: t.price ?? null,
      folio: h.folio ?? null,
      source: h.source,
      import_id: importId,
    });
  }
}

export function processParseResult(store: Store, result: ParseResult): ImportBatch {
  const batch: ImportBatch = {
    id: store.nextId("imports"),
    source: result.source,
    file_name: result.file_name,
    status: result.hasError() ? "failed" : "completed",
    count_parsed: result.holdings.length,
    count_imported: 0, count_merged: 0, count_duplicate: 0, count_skipped: 0, count_unclassified: 0,
    diagnostics: JSON.stringify(result.diagnostics),
    created_at: new Date().toISOString(),
  };
  store.imports.push(batch);

  const index = loadOverrideIndex(store);
  let imported = 0, merged = 0, duplicate = 0, unclassified = 0;
  const skipped = result.diagnostics.filter((d) => d.message.includes("Skipped")).length;

  for (const h of result.holdings) {
    const account = findOrCreateAccount(store, h);
    const [instrument, existed] = findOrCreateInstrument(store, h);
    const outcome = upsertHolding(store, account, instrument, h, batch.id);
    if (outcome === "duplicate") duplicate++;
    else {
      imported++;
      if (existed) merged++;
    }
    persistTransactions(store, account.id, instrument.id, h, batch.id);
    const cls = classifyInstrument(store, instrument, index);
    if (cls.asset_class === AssetClass.UNCLASSIFIED) unclassified++;
  }

  batch.count_imported = imported;
  batch.count_merged = merged;
  batch.count_duplicate = duplicate;
  batch.count_skipped = skipped;
  batch.count_unclassified = unclassified;

  const pruned = pruneUnbackedTradebookHoldings(store);
  if (pruned > 0) {
    const diags = JSON.parse(batch.diagnostics || "[]");
    diags.push({
      level: "info",
      message: `Excluded ${pruned} tradebook position(s) not present in your current holdings — likely sold, transferred out, or affected by a split/bonus. Their trades are kept for history; they don't count toward net worth.`,
      context: null,
    });
    batch.diagnostics = JSON.stringify(diags);
  }
  return batch;
}
